cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkWriterContext.java

```diff
@@ -21,7 +21,9 @@
 
 import java.io.Serializable;
 
-public interface BulkWriterContext extends Serializable
+import org.apache.cassandra.spark.common.JobStats;
+
+public interface BulkWriterContext extends Serializable, JobStats
```

Comment:
Rather than extending JobStats, please add a `jobStats` method to this interface and return an instance of something that implements `JobStats` like the other methods here do. We did this originally because the number of things this interface included was huge, so we decided to break them up into smaller implementation classes.

An example `JobStatsImpl` class might look something like this:
```java
package org.apache.cassandra.spark.bulkwriter;

import java.util.HashMap;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.apache.cassandra.spark.common.JobStats;

class JobStatsImpl implements JobStats
{
    private static final Logger LOGGER = LoggerFactory.getLogger(JobStatsImpl.class);;
    Map<String, String> stats = new HashMap<>();

    public JobStatsImpl(String sparkVersion, JobInfo jobInfo)
    {
        stats.put("sparkVersion", sparkVersion);
        stats.put("jobId", jobInfo.getId().toString());
        stats.put("keyspace", jobInfo.getId().toString());
        stats.put("table", jobInfo.getId().toString());
    }

    public void record(Map<String, String> statsToRecord)
    {
        stats.putAll(statsToRecord);
    }

    public void publish()
    {
        LOGGER.info("Job Stats: {}", stats);
    }
}
```

Then, CassandraBulkWriterContext just creates a new one of these in its constructor and returns it from the `jobStats` method. The actual field can/should be marked `transient` as Francisco has pointed out, so it doesn't get transferred to executors (and, therefore, it _doesn't_ need to be Serializable).

Reply:
Makes sense. Incorporated this change